### ISSUE TITLE
Prevent the use of the mailinabox.email website as an attack vector

### ISSUE THREAD

**Issue by JoshData** [mail-in-a-box/mailinabox#250]
The update instructions say to `curl ... | bash` a remote script. This is ok for first-time installations but it could be used as an attack vector to compromise machines as they update. Instead, the box should:
1. ship its own update script
2. know the gpg key I use to sign the git tags
3. don't update unless the tag is properly signed

**Comment by mlissner** [comment]
:+1:

**Comment by nstanke** [comment]
Before I submit an pull request, my attempt to fix this issue. 724c5c59398a0f0012be0a33de806bb758479932

Uses can run `/etc/init.d/mailinabox update` to install the latest signed tag. If we would log the current MiaB version, or the local git repository location (other than `/usr/local/bin/mailinabox`) we could skip the process when we're up-to-date.

Also I would suggest to move `/usr/local/bin/mailinabox` to the init script, so uses can just run `mailinabox update` to update their box.

We would still need a (secure) way to get @JoshData's PGP key on their box. Shipping it with `bootstrap.sh` would still leave `mailinabox.email` as an possible attack vector on first installs.

Let me know if you had something else in mind.

-Norman

**Comment by JoshData** [comment]
I tried something like this a while back:

https://github.com/mail-in-a-box/mailinabox/commit/47b217ce965d0ad249e26995f569afb8f8b79978
(not on master)

I'm not sure the init.d script is the right place for this. I also left some code comments. And it may be problematic with the way the bootstrap script does a shallow clone (see my commit).

**Comment by nstanke** [comment]
I'm sorry, didn't see your commit. We should use your solution then.

-Norman

**Comment by JoshData** [comment]
There was some reason I stopped working on it but I'm not sure now. Could you try it out and see if it even works?

**Comment by nstanke** [comment]
Committed some changes to my branch (https://github.com/nstanke/mailinabox/commits/update) . I'll test your commit tomorrow and let you know.

**Comment by nstanke** [comment]
Hi there,

I couldn't get your commit 47b217c to work. Mail-in-a-Box gets updated even when verification fails, also there is a (testing?) `exit 0` on line 60. 

Didn't face a problem with my "version" yet.

**Comment by JoshData** [comment]
> Mail-in-a-Box gets updated even when verification fails

Maybe piping it to grep wasn't a good idea.

> also there is a (testing?) exit 0 on line 60.

Ah yes.

**Comment by nstanke** [comment]
Updated my branch (9dfe7bf77ec3234eff9e149716ffd2bb15cf44bb) to use the new API endpoints. Would love to see some kind of build-in update mechanism.

**Comment by OmgImAlexis** [comment]
@nstanke about the "secure PGP key issue", @JoshData could setup a keybase account just for mail-in-the-box or one from himself and use the PGP key from there. Incase it ever got compromised he could even revoke it and issue a new one meaning it wouldn't update to anything signed with the old compromised key?

**Comment by JoshData** [comment]
Fyi, here's what I have so far: https://github.com/mail-in-a-box/mailinabox/commits/secureupgrade

I was going to add @nstanke's idea of using the new api to check for the latest version too.

**Comment by JoshData** [comment]
I was also thinking that shipping Mail-in-a-Box as a .deb through Launchpad could solve these issues.

**Comment by nstanke** [comment]
Would be definitely a cool & user-friendly way to ship Mail-in-a-Box through Launchpad, what happened to #183?

**Comment by allo-** [comment]
Ah, sorry for the duplicate.

The second issue from #566: It seems you can NEVER trust copy&paste. something easy to type would be more secure. http://thejh.net/misc/website-terminal-copy-paste.
This is not the most dangerous one, as you propably see for a short moment, that you're pasting much more than expected. (A Beginner may not know what to expect).

A nice idea for a bit more security is in the source code of https://hashbang.sh/

**Comment by JoshData** [comment]
> It seems you can NEVER trust copy&paste.

No, look, there are lots of ways that things can go wrong. But it doesn't help to speak in terms of absolutes ("never trust"). Whether what you copy & paste from the website is actually the simple `curl | sh` command or not is the least of your worries. You're about to execute code you haven't seen as root anyway!

hashbang.sh seems to suggest PGP-signing the bash script. I already do this -- see the README at the root of this repository for details.

**Comment by Hooloovoo** [comment]
> I was also thinking that shipping Mail-in-a-Box as a .deb through Launchpad could solve these issues.
As mentioned in #183 just now, I suspect these days a Snap would be better than a .deb in a PPA (and has the added advantage that it addresses the LTS upgrade problem). Hopefully a bit easier to create and manage, too, though I have not yet had a chance to proof-of-concept it.
